File: Micro-Services/pong_project/pong_project/static/js/pong/physics.js
export class Vector {
	constructor(x, y) {
		this.x = x;
		this.y = y;
	}
}

export class Rectangle {
	constructor(px, py, sx, sy, vx, vy) {
		this.size = new Vector(sx, sy);
		this.position = new Vector(px, py);
		this.velocity = new Vector(vx, vy);
	}
}

export class Collision {
	constructor(t, x, y, nx, ny) {
		this.time = t;
		this.point = new Vector(x, y);
		this.normal = new Vector(nx, ny);
	}
}

export class Particle {
	constructor(px, py, sx, sy, speed, lifetime) {
		this.life = lifetime;
		this.speed = speed;
		this.active = false;
		this.rectangle = new Rectangle(px, py, sx, sy, (Math.random() - 0.5) * speed, (Math.random() - 0.5) * speed);
	}
}

export class ParticlePool {
	constructor(size) {
		this.size = size;
		this.particle_lifetime = 1.5;
		this.pool = Array.from({ length: size }, () => new Particle(0, 0, 4, 4, 100, this.particle_lifetime));
	}

	get() {
		return this.pool.filter(particle => particle.active)
	}

	get_n_actives() {
		let actives = 0;
		for (let particle of this.pool) {
			if (particle.active)
				actives++;
		}
		return actives;
	}

	reset(x, y) {
		for (let particle of this.pool) {
			particle.active = true;
			particle.life = this.particle_lifetime;
			particle.rectangle.position.x = x;
			particle.rectangle.position.y = y;
			particle.rectangle.velocity.x = (Math.random() - 0.5) * particle.speed;
			particle.rectangle.velocity.y = (Math.random() - 0.5) * particle.speed;
		}
	}

	update(dt) {
		for (let particle of this.pool) {
			if (particle.active) {
				particle.rectangle.position.x += particle.rectangle.velocity.x * dt;
				particle.rectangle.position.y += particle.rectangle.velocity.y * dt;
				particle.life -= dt;
				if (particle.life <= 0)
					particle.active = false;
			}
		}
	}
}

function aabb_collision_test(r1, r2) {
	return r1.position.x + r1.size.x >= r2.position.x &&
		r1.position.x <= r2.position.x + r2.size.x &&
		r1.position.y + r1.size.y >= r2.position.y &&
		r1.position.y <= r2.position.y + r2.size.y;
}

function aabb_collision_resolve(r1, r2) {
	const buffer = 1;
	const overlap_x = Math.min(r1.position.x + r1.size.x - r2.position.x, r2.position.x + r2.size.x - r1.position.x);
	const overlap_y = Math.min(r1.position.y + r1.size.y - r2.position.y, r2.position.y + r2.size.y - r1.position.y);

	if (overlap_x < overlap_y) {
		if (r1.position.x < r2.position.x) {
			r1.position.x -= overlap_x + buffer;
		} else {
			r1.position.x += overlap_x + buffer;
		}
	} else {
		if (r1.position.y < r2.position.y) {
			r1.position.y -= overlap_y + buffer;
		} else {
			r1.position.y += overlap_y + buffer;
		}
	}
}


/**
 * Checks if a ray intersects with a rectangle.
 *
 * @param {Vector} origin - The origin of the ray.
 * @param {Vector} direction - The direction of the ray.
 * @param {Rectangle} target - The rectangle to test against
 * @returns {Collision} - A Collision object containing all the
 * informations about the collision. If no collision what detected, the 't'
 * parameter of the object is set to -1.
 */
export function ray_rectangle_collision(origin, direction, target) {
	let t_near = new Vector(
		direction.x === 0 ? (target.position.x - origin.x) > 0 ? Infinity : -Infinity : (target.position.x - origin.x) / direction.x,
		direction.y === 0 ? (target.position.y - origin.y) > 0 ? Infinity : -Infinity : (target.position.y - origin.y) / direction.y
	);
	let t_far = new Vector(
		direction.x === 0 ? (target.position.x + target.size.x - origin.x) > 0 ? Infinity : -Infinity : (target.position.x + target.size.x - origin.x) / direction.x,
		direction.y === 0 ? (target.position.y + target.size.y - origin.y) > 0 ? Infinity : -Infinity : (target.position.y + target.size.y - origin.y) / direction.y
	);

	if (t_near.x > t_far.x) [t_near.x, t_far.x] = [t_far.x, t_near.x];
	if (t_near.y > t_far.y) [t_near.y, t_far.y] = [t_far.y, t_near.y];

	let collision = new Collision(-1, 0, 0, 0, 0);

	if (t_near.x > t_far.y || t_near.y > t_far.x) return collision;

	let t_hit_near = Math.max(t_near.x, t_near.y);
	let t_hit_far = Math.min(t_far.x, t_far.y);

	if (t_hit_far < 0) return collision;

	collision.time = t_hit_near;
	collision.point.x = origin.x + t_hit_near * direction.x;
	collision.point.y = origin.y + t_hit_near * direction.y;

	if (t_near.x > t_near.y) {
		if (direction.x < 0) { collision.normal.x = 1; collision.normal.y = 0; }
		else { collision.normal.x = -1; collision.normal.y = 0; }
	} else if (t_near.x < t_near.y) {
		if (direction.y < 0) { collision.normal.x = 0; collision.normal.y = 1; }
		else { collision.normal.x = 0; collision.normal.y = -1; }
	}
	return collision;
}

/**
 * Update the velocity of r1 to bring it in contact with the object the ray
 * collided with.
 *
 * @param {Rectangle} r1 - The moving rectangle.
 * @param {Collision} collision - An object containing the collision time and normal.
 */
export function aabb_continuous_resolve(r1, collision) {
	return new Vector(
		r1.velocity.x + collision.normal.x * Math.abs(r1.velocity.x) * (1 - collision.time),
		r1.velocity.y + collision.normal.y * Math.abs(r1.velocity.y) * (1 - collision.time)
	);
}

/**
 * Test one moving rectangle against one static rectangle for collision.
 *
 * @param {Rectangle} r1 - The moving rectangle to test for collision.
 * @param {Rectangle} r2 - The static rectangle to test against.
 * @param {float} dt
 * @returns {Collision} - A Collision object containing all the
 * informations about the collision. If no collision what detected, the 't'
 * parameter of the object is set to -1.
 */
export function aabb_continuous_detection(r1, r2, dt) {

	if (aabb_collision_test(r1, r2)) {
		aabb_collision_resolve(r1, r2);
	}

	/* If r1 is not moving, a collision cannot occur */
	if (r1.velocity.x == 0 && r1.velocity.y == 0)
		return new Collision(-1, 0, 0, 0, 0);

	/* Create an expanded target to check against so that we can detect and resolve the collision properly */
	let r2_expanded = new Rectangle(
		r2.position.x - r1.size.x / 2,
		r2.position.y - r1.size.y / 2,
		r2.size.x + r1.size.x,
		r2.size.y + r1.size.y,
		0, 0
	);
	let collision = ray_rectangle_collision(
		new Vector(r1.position.x + r1.size.x / 2, r1.position.y + r1.size.y / 2),
		new Vector(r1.velocity.x * dt, r1.velocity.y * dt),
		r2_expanded
	);

	return collision;
}

File: Micro-Services/game_app/game_app/pong/physics.py
import random

class Vector:
    def __init__(self, x: float, y: float):
        self.x = x
        self.y = y


class Rectangle:
    def __init__(self, px: float, py: float, sx: float, sy: float, vx: float, vy: float):
        self.size = Vector(sx, sy)
        self.position = Vector(px, py)
        self.velocity = Vector(vx, vy)

class Collision:
    def __init__(self, t: float, x: float, y: float, nx: float, ny: float):
        self.time = t
        self.point = Vector(x, y)
        self.normal = Vector(nx, ny)


class Particle:
    def __init__(self, px, py, sx, sy, speed, lifetime):
        self.life = lifetime
        self.speed = speed
        self.active = False
        self.rectangle = Rectangle(px, py, sx, sy, (random.random() - 0.5) * speed, (random.random() - 0.5) * speed)


class ParticlePool:
    def __init__(self, size):
        self.size = size
        self.particle_lifetime = 1.5
        self.pool = [Particle(0, 0, 4, 4, 100, self.particle_lifetime) for _ in range(size)]

    def get(self):
        active_particles = [particle for particle in self.pool if particle.active]
        return active_particles

    def get_n_actives(self):
        return sum(1 for particle in self.pool if particle.active)

    def reset(self, x, y):
        for particle in self.pool:
            particle.active = True
            particle.life = self.particle_lifetime
            particle.rectangle.position.x = x
            particle.rectangle.position.y = y
            particle.rectangle.velocity.x = (random.random() - 0.5) * particle.speed
            particle.rectangle.velocity.y = (random.random() - 0.5) * particle.speed

    def update(self, dt):
        for particle in self.pool:
            if particle.active:
                particle.rectangle.position.x += particle.rectangle.velocity.x * dt
                particle.rectangle.position.y += particle.rectangle.velocity.y * dt
                particle.life -= dt
                if particle.life <= 0:
                    particle.active = False


def ray_rectangle_collision(origin: Vector, direction: Vector, target: Rectangle) -> Collision:
    t_near = Vector(
        (
            (target.position.x - origin.x) / direction.x
            if direction.x != 0
            else (float("inf") if (target.position.x - origin.x) > 0 else float("-inf"))
        ),
        (
            (target.position.y - origin.y) / direction.y
            if direction.y != 0
            else (float("inf") if (target.position.y - origin.y) > 0 else float("-inf"))
        ),
    )
    t_far = Vector(
        (
            (target.position.x + target.size.x - origin.x) / direction.x
            if direction.x != 0
            else (float("inf") if (target.position.x + target.size.x - origin.x) > 0 else float("-inf"))
        ),
        (
            (target.position.y + target.size.y - origin.y) / direction.y
            if direction.y != 0
            else (float("inf") if (target.position.y + target.size.y - origin.y) > 0 else float("-inf"))
        ),
    )

    if t_near.x > t_far.x:
        t_near.x, t_far.x = t_far.x, t_near.x
    if t_near.y > t_far.y:
        t_near.y, t_far.y = t_far.y, t_near.y

    collision = Collision(-1, 0, 0, 0, 0)

    if t_near.x > t_far.y or t_near.y > t_far.x:
        return collision

    t_hit_near = max(t_near.x, t_near.y)
    t_hit_far = min(t_far.x, t_far.y)

    if t_hit_far < 0:
        return collision

    collision.time = t_hit_near
    collision.point.x = origin.x + t_hit_near * direction.x
    collision.point.y = origin.y + t_hit_near * direction.y

    if t_near.x > t_near.y:
        if direction.x < 0:
            collision.normal.x = 1
            collision.normal.y = 0
        else:
            collision.normal.x = -1
            collision.normal.y = 0
    elif t_near.x < t_near.y:
        if direction.y < 0:
            collision.normal.x = 0
            collision.normal.y = 1
        else:
            collision.normal.x = 0
            collision.normal.y = -1

    return collision


def aabb_collision_test(r1, r2):
    return (r1.position.x + r1.size.x >= r2.position.x and
            r1.position.x <= r2.position.x + r2.size.x and
            r1.position.y + r1.size.y >= r2.position.y and
            r1.position.y <= r2.position.y + r2.size.y)

def aabb_collision_resolve(r1, r2):
    buffer = 1
    overlap_x = min(r1.position.x + r1.size.x - r2.position.x,
                    r2.position.x + r2.size.x - r1.position.x)
    overlap_y = min(r1.position.y + r1.size.y - r2.position.y,
                    r2.position.y + r2.size.y - r1.position.y)

    if overlap_x < overlap_y:
        if r1.position.x < r2.position.x:
            r1.position.x -= overlap_x + buffer
        else:
            r1.position.x += overlap_x + buffer
    else:
        if r1.position.y < r2.position.y:
            r1.position.y -= overlap_y + buffer
        else:
            r1.position.y += overlap_y + buffer

def aabb_continuous_resolve(r1: Rectangle, collision: Collision) -> None:
    return Vector(
        r1.velocity.x + collision.normal.x * abs(r1.velocity.x) * (1 - collision.time),
        r1.velocity.y + collision.normal.y * abs(r1.velocity.y) * (1 - collision.time),
    )


def aabb_continuous_detection(r1: Rectangle, r2: Rectangle, dt: float) -> Collision:

    if aabb_collision_test(r1, r2):
        aabb_collision_resolve(r1, r2)

    if r1.velocity.x == 0 and r1.velocity.y == 0:
        return Collision(-1, 0, 0, 0, 0)

    r2_expanded = Rectangle(
        r2.position.x - r1.size.x / 2,
        r2.position.y - r1.size.y / 2,
        r2.size.x + r1.size.x,
        r2.size.y + r1.size.y,
        0,
        0,
    )
    collision = ray_rectangle_collision(
        Vector(r1.position.x + r1.size.x / 2, r1.position.y + r1.size.y / 2),
        Vector(r1.velocity.x * dt, r1.velocity.y * dt),
        r2_expanded,
    )

    return collision
